src/Symfony/Component/EventDispatcher/CompiledEventDispatcher.php

```diff
@@ -0,0 +1,241 @@
+     */
+    public function __construct(IntrospectableContainerInterface $container, array $listeners = array())
+    {
+        $this->container = $container;
+        $this->listeners = $listeners;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function dispatch($eventName, Event $event = null)
+    {
+        if ($event === NULL) {
+            $event = new Event();
+        }
+
+        $event->setDispatcher($this);
+        $event->setName($eventName);
+
+        if (isset($this->listeners[$eventName])) {
+            // Sort listeners if necessary.
+            if (isset($this->unsorted[$eventName])) {
+                krsort($this->listeners[$eventName]);
+                unset($this->unsorted[$eventName]);
+            }
+
+            // Invoke listeners and resolve callables if necessary.
+            foreach ($this->listeners[$eventName] as &$definitions) {
+                foreach ($definitions as $key => &$definition) {
+                    if (!isset($definition['callable'])) {
+                        $definition['callable'] = array($this->container->get($definition['service'][0]), $definition['service'][1]);
+                    }
+
+                    call_user_func($definition['callable'], $event, $eventName, $this);
+                    if ($event->isPropagationStopped()) {
+                        return $event;
+                    }
+                }
+            }
+        }
+
+        return $event;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getListeners($eventName = null)
+    {
+        $result = array();
+
+        if ($eventName === NULL) {
+            // If event name was omitted, collect all listeners of all events.
+            foreach (array_keys($this->listeners) as $eventName) {
+                $listeners = $this->getListeners($eventName);
+                if (!empty($listeners)) {
+                    $result[$eventName] = $listeners;
+                }
+            }
+        } elseif (isset($this->listeners[$eventName])) {
+            // Sort listeners if necessary.
+            if (isset($this->unsorted[$eventName])) {
+                krsort($this->listeners[$eventName]);
+                unset($this->unsorted[$eventName]);
+            }
+
+            // Collect listeners and resolve callables if necessary.
+            foreach ($this->listeners[$eventName] as &$definitions) {
+                foreach ($definitions as $key => &$definition) {
+                    if (!isset($definition['callable'])) {
+                        $definition['callable'] = array($this->container->get($definition['service'][0]), $definition['service'][1]);
+                    }
+
+                    $result[] = $definition['callable'];
+                }
+            }
+        }
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function hasListeners($eventName = null)
+    {
+        return (bool) count($this->getListeners($eventName));
```

Comment:
This does not work because `$this->listeners[$eventName]` contains an array (keyed by priority) of arrays (the actual definitions). After adding and then removing a listener, `$this->listeners[$eventName]` will still contain entries, despite the fact that there are no listeners anymore. Also as long as there are no services to resolve, `getListeners` is cheap. I do not expect `hasListeners` being used exclusively (i.e. without a call to `getListeners` or `dispatch`), thus this optimization wouldn't bring too much anyway.